extension/js/common/api/email-provider/gmail/google-auth.ts

```diff
@@ -200,72 +189,39 @@ export class GoogleAuth {
-    return title.match(/^(?:https?:\/\/)?accounts\.google\.com/) !== null
-      || title.startsWith(GOOGLE_OAUTH_SCREEN_HOST.replace(/^https?:\/\//, ''))
-      || title.startsWith(GOOGLE_OAUTH_SCREEN_HOST);
-  };
-
-  private static isForwarding = (title: string) => {
-    return title.match(/^Forwarding /) !== null;
-  };
-
-  private static waitForAndProcessOauthWindowResult = async (windowId: number, acctEmail: string | undefined, scopes: string[], csrfToken: string, save: boolean): Promise<AuthRes> => {
-    while (true) {
-      const [oauth] = await tabsQuery({ windowId });
-      if (oauth?.title && oauth.title.includes(GoogleAuth.OAUTH.state_header) && !GoogleAuth.isUrl(oauth.title) && !GoogleAuth.isForwarding(oauth.title)) {
-        const { result, error, code, csrf } = GoogleAuth.processOauthResTitle(oauth.title);
-        if (error === 'access_denied') {
-          return { acctEmail, result: 'Denied', error, id_token: undefined }; // sometimes it was coming in as {"result":"Error","error":"access_denied"}
-        }
-        if (result === 'Success') {
-          if (!csrf || csrf !== csrfToken) {
-            return { acctEmail, result: 'Error', error: `Wrong oauth CSRF token. Please try again.`, id_token: undefined };
-          }
-          if (code) {
-            const { id_token } = save ? await GoogleAuth.retrieveAndSaveAuthToken(code, scopes) : await GoogleAuth.googleAuthGetTokens(code);
-            const { email } = GoogleAuth.parseIdToken(id_token);
-            if (!email) {
-              throw new Error('Missing email address in id_token');
-            }
-            return { acctEmail: email, result: 'Success', id_token };
-          }
-          return { acctEmail, result: 'Error', error: `Google auth result was 'Success' but no auth code`, id_token: undefined };
+      const responseUrl = await GoogleAuth.oauthLogin(authUrl);
+      if (!responseUrl) {
+        return { acctEmail, result: 'Denied', error: 'Invalid response url', id_token: undefined };
+      }
+      const url = new URL(responseUrl!);
+      const allowedScopes = this.getParameterByName(url.search, 'scope');
+      const code = this.getParameterByName(url.search, 'code') ?? '';
+      if (!allowedScopes?.includes(this.OAUTH.scopes.compose) || !allowedScopes?.includes(this.OAUTH.scopes.modify)) {
+        if (code !== '') {
+          // Try to get auth token to let login authorization be granted
+          await GoogleAuth.googleAuthGetTokens(code);
         }
-        return { acctEmail, result, error: error ? error : '(no error provided)', id_token: undefined };
       }
-      await Ui.time.sleep(250);
+      if (!code) {
+        return { acctEmail, result: 'Denied', error: "Google auth result was 'Success' but no auth code", id_token: undefined };
+      }
+      const { id_token } = save ? await GoogleAuth.retrieveAndSaveAuthToken(code, allowedScopes?.split(' ') ?? []) : await GoogleAuth.googleAuthGetTokens(code);
```

Comment:
In how many situations is this `save` used vs where is it not used? To try to have method with fewer disparate responsibilities, it may be worthwhile to remove it from here and explicitly save it where needed. But this depends how many places are we talking about. Since this is a private method, maybe it's just one place it's being called from? Then you could add the `if / save` there.

Reply:
IMHO I think better to leave this as it is for now.
Only one place has `save` as false and as we also  save `scopes` in `googleAuthSaveTokens`, better to refactor this once we migrate to `load allowed scopes from google api` feature.

https://github.com/FlowCrypt/flowcrypt-browser/blob/37d466bf471bde88eae19a2ccd17f1bf74ce87d2/extension/js/common/api/email-provider/gmail/google-auth.ts#L295

![image](https://user-images.githubusercontent.com/77344191/189095076-06267b8e-169b-41f0-ae90-523ae8e1d9a7.png)

Reply:
Ok, please link this there to discuss further. Now I wonder, do we really need to do `save: false` in that situation? Maybe it wouldn't hurt to save there also?

Reply:
Ok, Done.
I agree with you. but not sure exactly.